packages/react-reconciler/src/ReactFiberWorkLoop.js

```diff
@@ -1063,38 +1035,44 @@ function renderRoot(
         // Don't bother with a very short suspense time.
         if (msUntilTimeout > 10) {
           if (workInProgressRootHasPendingPing) {
-            // This render was pinged but we didn't get to restart earlier so try
-            // restarting now instead.
-            prepareFreshStack(root, expirationTime);
-            return renderRoot.bind(null, root, expirationTime);
+            const lastPingedTime = root.lastPingedTime;
+            if (lastPingedTime === NoWork || lastPingedTime >= expirationTime) {
+              // This render was pinged but we didn't get to restart earlier so
+              // try restarting now instead.
+              root.lastPingedTime = expirationTime;
+              prepareFreshStack(root, expirationTime);
+              return;
```

Comment:
If `lastPingedTime === expirationTime` then I wouldn't expect there to be a callback already scheduled, right? So how does this pick up work back up?

Reply:
https://github.com/facebook/react/pull/16715/files/6ace553825626774d9d9325be84389fe7fe5e33a..fd253c9b0662c48f9b780c6d48bc884c99cace7f#r322551325